### PR TITLE
Refactor TestSnapshot

### PR DESCRIPTION
**Pull request by shuheiktgw**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Refactored TestSnapshot and added some missing tests. 

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/issues/72

### CODE REVIEW

File: pkg/cache/snapshot_test.go [hunk 1]
```diff
@@ -40,149 +39,208 @@ func TestSnapshot(t *testing.T) {
...
 	}
-	cache := New(fake.NewClientBuilder().WithScheme(scheme).Build())
-	clusterQueues := []kueue.ClusterQueue{
+
+	cases := []struct {
+		name          string
+		clusterQueues []*kueue.ClusterQueue
+		flavors       []kueue.ResourceFlavor
+		workloads     []*kueue.Workload
+		wantSnapshot  func(workloads []*kueue.Workload) Snapshot
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:48]:
```suggestion
		wantSnapshot  Snapshot
```

Avoid building the desired results of the test.

File: pkg/cache/snapshot_test.go [hunk 2]
```diff
@@ -40,149 +39,208 @@ func TestSnapshot(t *testing.T) {
 	if err := kueue.AddToScheme(scheme); err != nil {
 		t.Fatalf("Failed adding kueue scheme: %s", err)
 	}
-	cache := New(fake.NewClientBuilder().WithScheme(scheme).Build())
-	clusterQueues := []kueue.ClusterQueue{
+
+	cases := []struct {
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:43]:
```suggestion
	cases := map[string]struct {
```

This way, the cases might run in a different order each time, which ensure that they don't depend on each other.

File: pkg/cache/snapshot_test.go [hunk 3]
```diff
@@ -40,149 +39,208 @@ func TestSnapshot(t *testing.T) {
...
+			wantSnapshot: func(workloads []*kueue.Workload) Snapshot {
+				return Snapshot{
+					ClusterQueues: map[string]*ClusterQueue{
+						"bar": {
+							Name:                 "bar",
+							NamespaceSelector:    labels.Everything(),
+							Status:               Active,
+							RequestableResources: map[corev1.ResourceName][]FlavorLimits{},
+							UsedResources:        Resources{},
+							Workloads:            map[string]*workload.Info{},
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:74]: you can replace this with `utiltesting.MakeClusterQueue` too.

File: pkg/cache/snapshot_test.go [hunk 4]
```diff
@@ -40,149 +39,208 @@ func TestSnapshot(t *testing.T) {
...
-								Name: "spot",
-								Quota: kueue.Quota{
-									Min: resource.MustParse("100"),
-								},
-							},
+			wantSnapshot: func(workloads []*kueue.Workload) Snapshot {
+				return Snapshot{
+					ClusterQueues:            map[string]*ClusterQueue{},
+					ResourceFlavors:          map[string]*kueue.ResourceFlavor{},
+					InactiveClusterQueueSets: sets.String.Insert(sets.NewString("flavor-nonexistent")),
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:103]:
```suggestion
					InactiveClusterQueueSets: sets.NewString("flavor-nonexistent"),
```

File: pkg/cache/snapshot_test.go [hunk 5]
```diff
@@ -40,149 +39,208 @@ func TestSnapshot(t *testing.T) {
...
+				}
+			},
+		},
+		{
+			name: "resource flavors",
+			flavors: []kueue.ResourceFlavor{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "demand"},
+					Labels:     map[string]string{"foo": "bar", "instance": "demand"},
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:112]: Use `utiltesting`

File: pkg/cache/snapshot_test.go [hunk 6]
```diff
@@ -245,148 +293,157 @@ func TestSnapshot(t *testing.T) {
...
+			for i := range tc.flavors {
+				cache.AddOrUpdateResourceFlavor(&tc.flavors[i])
+			}
+
+			for _, w := range tc.workloads {
+				cache.AddOrUpdateWorkload(w.DeepCopy())
+			}
+
+			snapshot := cache.Snapshot()
+			if diff := cmp.Diff(tc.wantSnapshot(tc.workloads), snapshot, cmpopts.IgnoreUnexported(Cohort{})); diff != "" {
```
Comment (alculquicondor) [inline pkg/cache/snapshot_test.go:444]: Most of the things in Snapshot are just copies of the original objects. We probably care more about the accumulated resources in the cohort and we care less about all the fields in ClusterQueue or Workloads to match.

Maybe we can consider using `cmpopts.IgnoreTypes(WorkloadSpec{}, ClusterQueueSpec{})` to reduce the number of lines. WDYT?

**Comment by tenzen-y** [comment]
I take this over.
/assign

**Comment by alculquicondor** [comment]
/close
in favor of #538

**Pull request closed without merging**